src/txmempool.cpp

```diff
@@ -57,48 +57,69 @@ size_t CTxMemPoolEntry::GetTxSize() const
 // Update the given tx for any in-mempool descendants.
 // Assumes that setMemPoolChildren is correct for the given tx and all
 // descendants.
-void CTxMemPool::UpdateForDescendants(txiter updateIt, cacheMap &cachedDescendants, const std::set<uint256> &setExclude)
+void CTxMemPool::UpdateForDescendants(txiter update_it, cacheMap& cache, const std::set<uint256>& exclude)
 {
-    setEntries stageEntries, setAllDescendants;
-    stageEntries = GetMemPoolChildren(updateIt);
-
-    while (!stageEntries.empty()) {
-        const txiter cit = *stageEntries.begin();
-        setAllDescendants.insert(cit);
-        stageEntries.erase(cit);
-        const setEntries &setChildren = GetMemPoolChildren(cit);
-        for (txiter childEntry : setChildren) {
-            cacheMap::iterator cacheIt = cachedDescendants.find(childEntry);
-            if (cacheIt != cachedDescendants.end()) {
-                // We've already calculated this one, just add the entries for this set
-                // but don't traverse again.
-                for (txiter cacheEntry : cacheIt->second) {
-                    setAllDescendants.insert(cacheEntry);
+    const auto epoch = GetFreshEpoch();
+    const CTxMemPool::setEntries& direct_children = GetMemPoolChildren(update_it);
+    // set up the update_cache to contain all of our transaction's children (note --
+    // already de-duplicated in case multiple outputs of ours are spent in one
+    // transaction)
+    vecEntries update_cache;
+    update_cache.reserve(direct_children.size());
+    // mark every direct_child as visited so that we don't accidentally re-add them
+    // to the cache in the grandchild is child case
+    for (const txiter direct_child : direct_children) {
+        update_cache.emplace_back(direct_child);
+        visited(direct_child);
+    }
+    // already_traversed index keeps track of the elements that we've
+    // already expanded. If index is < already_traversed, we've walked it.
+    // If index is >= already_traversed, we need to walk it.
+    // If already_traversed >= update_cache.size(), we're finished.
+    for (size_t already_traversed = 0; already_traversed < update_cache.size(); /* modified in loop body */) {
+        // rotate the back() to behind already_traversed
+        const txiter child_it = update_cache.back();
+        std::swap(update_cache[already_traversed++], update_cache.back());
+
+        // N.B. grand_children may also be children
+        const CTxMemPool::setEntries& grand_children = GetMemPoolChildren(child_it);
+        for (const txiter grand_child_it : grand_children) {
+            if (visited(grand_child_it)) continue;
+            cacheMap::iterator cached_great_grand_children = cache.find(grand_child_it);
+            if (cached_great_grand_children != cache.end()) {
```

Comment:
From reading this code, it looks like if you have a hit in the cache, then you wouldn't end up adding `grand_child_it` to `update_cache`; you'd just add `grand_child_it`'s descendants?   That looks like a bug, though I haven't written a test to try to demonstrate it.

I know the cache goes away in the next commit, so assuming that makes sense then this may not be worth fixing (versus just squashing the two commits together), but I think it'd just be a two-line fix, if my understanding is correct?

Reply:
I think you're correct about that. The fix is to include grand_child_it and then swap it into the already_traversed index if there is a hit.

Reply:
fixed